### ISSUE TITLE
[Feature] support named parameters

### ISSUE THREAD

**Issue by AbdelrahmanHafez** [Automattic/mongoose#10367]
re https://github.com/Automattic/mongoose/pull/10360#issuecomment-862245274
It would be nice if we support named parameters for mongoose methods.
It should be straightforward to implement, I'll start with the following:
* Model.find
* Model.findOne
* Model.updateOne
* Model.updateMany
* Model.countDocuments
* Model.validate
* Model.aggregate
* Model.exists

Then we can decide what other functions we can apply the pattern to.

**Comment by swayamsingh2004** [comment]
I'd like to work on this. Since this covers multiple methods, I'll start by implementing named parameter support for Model.findOne as a first step, and we can decide on the rest after. Let me know if that scope works.

**Comment by AbdelrahmanHafez** [comment]
Hi @swayamsingh2004, I'll probably be slow to review, but will get to it eventually. If that works, your plan sounds great.

**Comment by swayamsingh2004** [comment]
While working on `Model.findOne`, I ran into a design question I'd like input on before writing code.

The challenge is detecting whether a call is using the old positional style or the new named-parameters style, since both pass a single object as the first argument in some cases:

```js
User.findOne({ email: "x" })                          // old style, plain filter
User.findOne({ filter: { email: "x" }, projection: "name" })  // new style, named params
```

A simple "does the object have a `filter` key" check breaks if someone's actual schema has a real field literally named `filter`:

```js
User.findOne({ filter: "some-value" })  // ambiguous: named-params wrapper, or a real filter on a `filter` field?
```

A couple of options I see:
1. Accept this as a known, documented limitation (collisions with a real `filter`/`projection`/`options` field are rare in practice).
2. Use a clearly distinct signature for the new style so it can never collide with a normal filter object (e.g. requiring a special marker, or a separate method name).

Wanted to check which direction you'd prefer before I commit to an implementation. If there's no strong preference, I'll go with option 1 as the default and move forward, just wanted to flag the tradeoff first.

**Comment by AbdelrahmanHafez** [comment]
We need to go with option 2, we generally support arbitrary schema paths unless they conflict with mongo server, should continue to support here for consistency and better support.

**Comment by swayamsingh2004** [comment]
Thanks for the clarification — that makes sense given the general philosophy of supporting arbitrary schema paths.

For option 2, I'd propose using a JavaScript `Symbol` as an internal marker to identify the named-parameters style, since Symbols can't collide with any string-based field name a user's schema might have, no matter what it's called (including `filter`, `projection`, etc.):

```js
// exported internally by mongoose
const namedParamsMarker = Symbol('mongoose#namedParams');

// detection inside Model.findOne
if (conditions != null && conditions[namedParamsMarker]) {
  // treat as named-parameters style
} else {
  // treat as old positional style, regardless of what keys it has
}
```

This guarantees zero collision with real schema fields, since a string key can never equal a Symbol key.

One thing I want to check before building this out: are Symbols safe to use here, or could they cause issues with serialization/BSON encoding somewhere deeper in the query pipeline? Wanted to confirm before going further with the implementation.

**Comment by AbdelrahmanHafez** [comment]
Thanks for picking this up @swayamsingh2004. I've been thinking about the `namedParamsSymbol` approach and I'd like to push back on it a bit before we build more methods on top of it.

The Symbol disambiguates perfectly, and to be fair it is typable (declare it as a `unique symbol` and put it in an interface, I checked). But the per-call ergonomics aren't great. The symbol has to be exported from our public types (it's currently internal in `lib/helpers/symbols.js`), and on every call the user imports it and writes `{ [namedParamsSymbol]: true, ... }`, which is noisy and doesn't autocomplete the way a normal property does.

I want to reconsider using `$`-prefixed keys instead (`$filter`, `$projection`, `$options`, `$update`, etc.):

```js
User.findOne({ $filter: { email: 'x@y.com' }, $projection: 'name', $options: { lean: true } })
```

The original objection was that a real schema field could be named `filter`, so plain keys are ambiguous. The `$` prefix sidesteps that: the only question that matters is whether the sentinel can appear at the top level of the positional argument it replaces, and for our methods it can't.

- Filter position (`find`, `findOne`, `countDocuments`, `deleteOne`, `updateOne`, `findOneAndUpdate`, etc.): the top-level query operators are a closed set (`$and $or $nor $text $where $comment $expr $jsonSchema`). `$filter` isn't one of them, it's an aggregation expression operator only valid nested inside `$project`/`$addFields`. `find({ $filter: ... })` gets rejected by the server as an unknown top-level operator, so no real positional call collides.
- Pipeline position (`aggregate`): the positional arg is an array, the wrapper is an object, so it's distinguishable before you even look at keys. `aggregate({ $pipeline: [...] })` is unambiguous.
- Update position: a classic update doc's top-level keys are `$set $unset $inc $push $addToSet ...`, and an update aggregation pipeline is an array of stages. Neither contains `$update`. The update is never in the detection position anyway, detection runs on the filter, and the update is just read as `$update`, whose value can be a doc or a pipeline array, handled downstream exactly as today:

```js
User.updateOne({ $filter: { _id }, $update: [{ $set: { name: 'x' } }], $options: { upsert: true } })
```

The `$`-prefix also gives you normal autocomplete (you type `$f` and the editor completes `$filter` off the overload's param type) without importing anything.

The one set of names to avoid is the DBRef trio (`$ref`/`$id`/`$db`). I checked this against a real server (8.2.6), since `$id` is the natural sentinel for `findById`. Two findings: modern MongoDB (the 5.0+ relaxation) lets you store `$id`/`$ref`/`$db` at the top level of a document, they're not always nested in a DBRef field; and more importantly the query parser treats a top-level `$id`/`$ref`/`$db` as a normal field match rather than an operator, so `find({ $id: 'foo' })` actually matches documents that have a top-level `$id` field. Every other `$`-name I tried is rejected as an unknown top-level operator (BadValue):

```
$filter $update $pipeline $projection $options $replacement  -> REJECTED (BadValue)
$identifier $byId $_id                                        -> REJECTED (BadValue)
$id $ref $db                                                  -> ACCEPTED as a data-field match
```

So the safety rule is simple: a `$`-sentinel is safe iff the server rejects it as a top-level query key, which holds for everything except real operators and the DBRef trio. That means we don't have to skip the `findById*` variants, we just can't name the sentinel `$id`. `$_id` works (server rejects it like the rest) and reads honestly since findById is an `_id` match:

```js
User.findById({ $_id: someId, $projection: 'name', $options: { lean: true } })
```

The other thing I'm undecided on is aesthetics. `$filter` and `$options` happen to match real operator names (`$filter` aggregation, `$options` regex flags), even though there's no actual collision at the top level since they're never detection triggers there. So it reads a little odd, but `$update`/`$pipeline`/`$projection` are clean. I don't think it's a blocker but worth a second opinion.

To keep it airtight I'd define detection as: named-params mode iff exactly one argument is passed, it's a non-null non-array object, at least one of the method's sentinels is present, and every top-level key is one of that method's known sentinels (required plus optionals like `$projection`/`$options`, which can be omitted). So `{ $filter: {...} }` on its own is valid. Anything else throws a clear error instead of silently falling back: a non-sentinel key mixed in, a typo like `$fitler`, or a named-params-looking object passed alongside a second positional argument. Requiring a single argument matters since `findOne({ $filter: ... }, 'name')` is always a mistake, and it's better to throw than silently misread it. It also types cleanly as an overload.

@vkarpov15 @hasezoey what do you think, Symbol vs `$`-prefix? Also, are you still open to this feature in the first place, or has your thinking changed since the original discussion? Don't want anyone building it out if the answer's no.

**Comment by swayamsingh2004** [comment]
Thanks for the detailed breakdown — really appreciate the depth here, especially verifying the `$`-prefix safety against a live server rather than assuming it from docs.

The ergonomics argument makes sense to me. Having `$filter`/`$projection`/`$options` autocomplete naturally without an import is a real improvement over the Symbol approach, and your point about the server already rejecting non-operator `$`-prefixed top-level keys closes the collision concern cleanly — that's a stronger guarantee than I'd originally reasoned through.

On the DBRef trio edge case: avoiding `$id`/`$ref`/`$db` as sentinel names and using `$_id` for `findById` sounds like the right call given what you found.

I'm happy to wait for input from @vkarpov15 and @hasezoey before changing direction. If the team lands on the `$`-prefix approach, I can rework the `Model.findOne` implementation and tests to match the detection rule you outlined (single argument, non-null non-array object, every top-level key a recognized sentinel for that method, clear error on anything else). Let me know once there's alignment and I'll get started.

**Comment by hasezoey** [comment]
I personally have no preference to whether this is implemented or not.
The `$` prefixed way seems the most ergonomic to me (instead of symbols), though with that i am concerned that it may collide with mongodb's operators or even allowed fields (as already pointed), especially in the future (unless there is some kind of non-change guarantee?).

Also, having to use `$_id` over `$id` in `findById` may be confusing. Maybe just use `$filter`here too (but allow `$filter: id`) ?

**Comment by AbdelrahmanHafez** [comment]
>it may collide with mongodb's operators or even allowed fields (as already pointed), especially in the future

I checked the current mongo fields and the params we pass, none conflict except `$id`. Of course that could change in the future, if/when that happens I think we can handle it by doing schema top level checks for paths that might conflict with named params and warn users so they know the risk of conflict, with a flag to suppress the warning.

>Also, having to use `$_id` over `$id` in `findById` may be confusing. Maybe just use `$filter`here too (but allow `$filter: id`) ?

Using `$filter: id` as the name would deviate from what we'd be doing everywhere else, where we'd name the keys as the original param name. I'm leaning more towards `$_id` to stay consistent with the rest of the calls. I'm also open to supporting `$filter: id` as an alternative way while keeping `$_id` supported.

WDYT?

**Comment by vkarpov15** [comment]
Asking user to set `[namedParamsMarker]` themselves is not my favorite. An alternative option would be what we do with [`mongoose.trusted()`](https://mongoosejs.com/docs/api/mongoose.html#Mongoose.prototype.trusted()): create a function that sets the `[namedParamsMarker]`.

I'm also not a huge fan of using `$filter` etc. because 1) TypeScript confusion, 2) [MongoDB does support storing fields that start with `$`](https://www.mongodb.com/docs/manual/core/dot-dollar-considerations/) so `$filter` is technically a valid field name. It's one thing for Mongoose to prefix methods like `$getChanges()`, but disallowing filtering documents with `$filter` as a field name seems like an unnecessary restriction.

Another option to consider would be an explicit `QueryParams` class and a `namedQueryParams` method, so users could do the following:

```javascript
await Model.findOne(
  mongoose.queryParams({
    filter,
    projection,
    options
  })
)
```

That would be cleaner for TypeScript purposes. Although using symbol markers is typically better performance wise in TypeScript (fewer instantiations) so there's some tradeoffs to consider.

**Comment by AbdelrahmanHafez** [comment]
`$filter` is storable but isn't directly queryable. You can store a top-level `$filter` field (5.0+), but `find({ $filter: 'x' })` is rejected by the server as an operator, so it was never a working query. The only way to filter on such a field is `$expr`/`$getField`, which isn't a marker and keeps working. So using `$filter` as a marker takes away nothing you can do today.

The real restriction is only the DBRef stuff (`$ref`/`$id`/`$db`), which the server treats as data, so `find({ $id: 'foo' })` does match. That's why I'd use `$_id` for findById, not `$id`. It doesn't apply to `$filter`/`$update`/etc.

On `queryParams()`: I'd rather avoid the wrapper. A big reason I want named params is testing, mocking chainable query APIs is painful, and a plain object collapses that into something I can build and assert on directly. A nested function call puts that layer back, which is the thing I'm trying to get away from.

TS concern is fair, but I think the detection rule plus a discriminated overload covers most of it. Happy to dig into a specific case if you have one.

WDYT?

**Comment by vkarpov15** [comment]
`$_id` is fine, I think that's more idiomatic than `$id`.

For `queryParams()`, why would that make testing more difficult? That would still remove the need for chaining APIs (although you don't necessarily have to use chaining APIs either). You would just need to assert that the `queryParams()` return value (either class or POJO) is what you expect.

Presuming we go with the `$filter` approach, how would we handle `Model.find({ $options })`? Would that also be treated as named parameters or would we require explicitly setting `Model.find({ $filter: null, $options })`? Defining "object should have at least one of these keys" in TypeScript seems like it would be difficult and performance intensive.

**Comment by AbdelrahmanHafez** [comment]
I built a quick `findOne` PoC to put numbers on the TS performance concern in #16343

I also benchmarked `find({ $options })`: making "at least one of the keys" valid needs a union type, it uses 3-5x the instantiations of a single interface, and that compounds across every method. So I'd require `$filter` (pass `$filter: null` for the rare options-only call) and keep the type a plain interface. Runtime still accepts `$options` alone.

One more thing on the TS side: since `QueryFilter` is permissive, a plain added overload gives autocomplete but doesn't actually reject malformed named params (they fall through to the positional filter). Branding the positional overloads with `& { $filter?: never; $projection?: never; $options?: never }` fixes that, `$typo` or a missing `$filter` become compile errors, for ~120 extra instantiations, so basically free.

@swayamsingh2004 Thank you for your interest in working on this, this feature turned out to be more nuanced than I initially thought, and it'd be more efficient for someone internal to pick it up. I'm happy to give commit attribution if you're okay with that, although the commit content will be a bit different from what you initially created.

**Comment by swayamsingh2004** [comment]
Thank you for the kind offer, @AbdelrahmanHafez — commit attribution would 
mean a lot to me. I'm glad the PR helped surface the design considerations, 
and I learned a great deal following the discussion between you, @hasezoey, 
and @vkarpov15. Looking forward to seeing this land!

**Comment by vkarpov15** [comment]
Yeah I'd say requiring `$filter: null` or `$id: null` is reasonable.

However, I'd like to take a step back and think a little more about what problem we're trying to solve with named parameters? Is it for readability like the original comments from #10360 indicate? Easier assertions when testing? What do you think @AbdelrahmanHafez ?

**Comment by AbdelrahmanHafez** [comment]
Two reasons:
1. A nicer way to pass everything at one parameter without having to worry about ordering of parameters.
2. Easier testing and mocking, mocking chained APIs is verbose and fragile, re-ordering the chain requires updating the tests. Having methods take just one object with everything in it makes it easier to test.

**Comment by vkarpov15** [comment]
I'm not certain I understand (2). For one thing, named parameters won't prevent chaining. For another, you don't need to use chaining with positional args.

**Comment by AbdelrahmanHafez** [comment]
You're right, it won't prevent chaining, and positional args with an `options` object already avoids it. It wouldn't prevent chaining, the point is it should mitigate the need to chain, which is what makes the tests simpler.

The pain is mocking the chained form people actually write:

```js
// production
User.find({ _id: id }).populate({ path: 'posts', select: 'title' }).lean();

// test: fake the whole chainable surface, one stub per link
// if the order of chaining changes, these stubs need to change
const lean = sinon.stub().resolves(docs);
const populate = sinon.stub().returns({ lean });
sinon.stub(User, 'find').returns({ populate });
// assertions are split across find/populate/lean and coupled to the chain order
```

vs everything in one object:

```js
// production
// no need for `$projection: null`
User.find({ $filter: { _id: id }, $options: { lean: true, populate: { path: 'posts', select: 'title' } } });

// test: one stub, resolves directly, asserted in one shot
const find = sinon.stub(User, 'find').resolves(docs);
sinon.assert.calledWith(find, { $filter: { _id: id }, $options: { lean: true, populate: { path: 'posts', select: 'title' } } });
```

You'r also right that positional `find(filter, null, { lean, populate })` gets you the same single-stub testability. But in practice people chain because positional means remembering order and padding skipped args with `null`. Named params give the single-object call site without that friction, so the easily-mockable form is also the ergonomic one.
